### PR TITLE
Fix indis dual-read lix switch issue

### PR DESCRIPTION
**Pull request by brycezhongqing**
## 1. Background
For gcn-39955
After [indis-dual-read lix](https://trex.corp.linkedin.com/trex/test/2081511207/iteration/4770044) switched to observer-only mode, job-postings-mt was not able to make downstream calls to d2 services under a symlink cluster(This issue has been fixed by #956 . Then D2 team tried to switch dual-read mode [lix](https://trex.corp.linkedin.com/trex/test/2081511207/iteration/4770044) from `Observer-only` to `DUAL_READ`, but it failed for partial downstream services.

### how to reproduce this :
- create a new test lix: https://trex.corp.linkedin.com/trex/test/2081519814/iteration/4770028
- change lix name from si.indis.client_read_mode to indis_automation_test  in container
- job-postings-mt bump up container snapshot and test in local ei.
- mint build && mint build-cfg -f dev && mint deploy -f dev in job-postings-mt

So we found,  after Lix switch from `Observer-only` to `DUAL_READ`, it will take a few minutes to update for partial downstream services. The root cause is  `all downstream services will use global rate-limit to update their dual-read mode which may cause partial services being starved and take a long time to update. 

## 3. Solution
After discussing with @bohhyang . We create a map, key is to downstream service name and value is the relative rate limiter for each service. So rate limiter will be service granularity. 

## 4. Test
- For rest.li(pegasus) repo, `./scripts/local-release -s` to create a snapshot
- Bump up pegasus version in container and for container `mint build && mint snapshot && mint release`
- Bump up `pegasus version` and `container version` in `job-postings-mt` and then `mint build && mint build-cfg -f dev && mint deploy -f dev in job-postings-mt`
- Check the service dual read mode(it's currently in `DUAL_READ` mode)
```
curli --fabric ei-ltx1 --pretty-print "https://zhonchen-mn2.linkedin.biz:9607/jobPostingsMt/resources/jobPostings/2147622124?action=purchase" --data '{"owner":"urn:li:contract:92952029", "isFreemiumPromotion": true,
"jobPurchaseBillingParams":{"dailyBudget":{"currencyCode":"USD", "amount":"116"}, "unitCost":{"currencyCode":"USD","amount":"3"},
"costType":"COST_PER_VIEW"}, "viewer":"urn:li:system:0"}' --dv-auth=SELF
```
image
We found all the downstream services switch to `DUAL_READ` mode
- switch the lix from `DUAL_READ` to  `Observer-only`, after 5 minutes, call 
```
curli --fabric ei-ltx1 --pretty-print "https://zhonchen-mn2.linkedin.biz:9607/jobPostingsMt/resources/jobPostings/2147622124?action=purchase" --data '{"owner":"urn:li:contract:92952029", "isFreemiumPromotion": true,
"jobPurchaseBillingParams":{"dailyBudget":{"currencyCode":"USD", "amount":"116"}, "unitCost":{"currencyCode":"USD","amount":"3"},
"costType":"COST_PER_VIEW"}, "viewer":"urn:li:system:0"}' --dv-auth=SELF
```
We found the service is in `Observer-only`
image

- switch the lix from `Observer-only` to  `DUAL_READ`, after 5 minutes, call 
```
curli --fabric ei-ltx1 --pretty-print "https://zhonchen-mn2.linkedin.biz:9607/jobPostingsMt/resources/jobPostings/2147622124?action=purchase" --data '{"owner":"urn:li:contract:92952029", "isFreemiumPromotion": true,
"jobPurchaseBillingParams":{"dailyBudget":{"currencyCode":"USD", "amount":"116"}, "unitCost":{"currencyCode":"USD","amount":"3"},
"costType":"COST_PER_VIEW"}, "viewer":"urn:li:system:0"}' --dv-auth=SELF
```
We found all downstream services are  in `DUAL_READ`
image

### CODE REVIEW

File: d2/src/main/java/com/linkedin/d2/balancer/dualread/DualReadStateManager.java [hunk 1]
```diff
@@ -191,7 +191,14 @@ public void checkAndSwitchMode(String d2ServiceName)
 
     _executorService.execute(() ->
     {
-      boolean shouldCheck = _rateLimiter.tryAcquire();
+      if(d2ServiceName == null){
+        return;
+      }
```
Comment (PapaCharlie) [inline d2/src/main/java/com/linkedin/d2/balancer/dualread/DualReadStateManager.java:196]: You should move this check outside of `_executorService`, there's no point in scheduling this if it won't run :) 
Also, watch the formatting!

**Review by PapaCharlie — APPROVED** [review]

File: d2/src/main/java/com/linkedin/d2/balancer/dualread/DualReadStateManager.java (hunk 1 above)
Comment (brycezhongqing) [reply d2/src/main/java/com/linkedin/d2/balancer/dualread/DualReadStateManager.java]: Thanks. Updated~

**Comment by bohhyang** [comment]
> switch the lix from xxx to xxx, after 5 minutes, call

To confirm, is the 5-min waiting for lix update to be propagated to lix client?

Also, we should check for many downstream services' dual read modes instead of just "accountBalances". The goal is to ensure once the lix value is updated, all downstream services should see the updated value (with their individual rate limiters). So a solid test should be, after X mins (once the lix update is propagated to the lix client), hitting endpoint A, B, C, etc (nearly the same time, in different terminal tabs for example) which call different downstream services, all downstream services' dual read modes are updated.

File: d2/src/main/java/com/linkedin/d2/balancer/dualread/DualReadStateManager.java [hunk 2]
```diff
@@ -188,10 +188,15 @@ public void checkAndSwitchMode(String d2ServiceName)
       LOG.info("Dual read mode executor is shut down already. Skipping getting the latest dual read mode.");
       return;
     }
-
+    if (d2ServiceName == null)
```
Comment (bohhyang) [inline d2/src/main/java/com/linkedin/d2/balancer/dualread/DualReadStateManager.java:191]: when service name is null, we should the original behavior which check the global rate limiter and call updateGlobal as needed.

**Review by bohhyang — CHANGES_REQUESTED** [review]

**Comment by brycezhongqing** [comment]
> > switch the lix from xxx to xxx, after 5 minutes, call
> 
> To confirm, is the 5-min waiting for lix update to be propagated to lix client?
> 
> Also, we should check for many downstream services' dual read modes instead of just "accountBalances". The goal is to ensure once the lix value is updated, all downstream services should see the updated value (with their individual rate limiters). So a solid test should be, after X mins (once the lix update is propagated to the lix client), hitting endpoint A, B, C, etc (nearly the same time, in different terminal tabs for example) which call different downstream services, all downstream services' dual read modes are updated. You don't need to set breakpoints to show the updated value which will cause hanging/delay, but just show the app log which will log about the updated values.



> > switch the lix from xxx to xxx, after 5 minutes, call
> 
> To confirm, is the 5-min waiting for lix update to be propagated to lix client?
> 
> Also, we should check for many downstream services' dual read modes instead of just "accountBalances". The goal is to ensure once the lix value is updated, all downstream services should see the updated value (with their individual rate limiters). So a solid test should be, after X mins (once the lix update is propagated to the lix client), hitting endpoint A, B, C, etc (nearly the same time, in different terminal tabs for example) which call different downstream services, all downstream services' dual read modes are updated. You don't need to set breakpoints to show the updated value which will cause hanging/delay, but just show the app log which will log about the updated values.

1. Yes. The 5-min waiting is lix update
2. Yes. I have checked all downstream services, `accountBalances` is one of them. All the downstream services will be updated dual-read mode.

File: d2/src/main/java/com/linkedin/d2/balancer/dualread/DualReadStateManager.java (hunk 2 above)
Comment (brycezhongqing) [reply d2/src/main/java/com/linkedin/d2/balancer/dualread/DualReadStateManager.java]: Thanks for the comments. Updated

**Comment by bohhyang** [comment]
> 2. Yes. I have checked all downstream services, `accountBalances` is one of them. All the downstream services will be updated dual-read mode.

Could you post the app log (or a snippet) about the updated dual read mode to the Test Done section?

**Comment by brycezhongqing** [comment]
> > 2. Yes. I have checked all downstream services, `accountBalances` is one of them. All the downstream services will be updated dual-read mode.
> 
> Could you post the app log (or a snippet) about the updated dual read mode to the Test Done section?

Updated

**Review by bohhyang — APPROVED** [review]
lgtm!

**Comment by shivamgupta1** [comment]
lgtm

**Pull request merged**